### DISCUSSION TITLE
Normalization in NeuralNetworkReductor

### DISCUSSION THREAD

**Discussion by jacktang** [pymor/pymor#1992] [Q&A]
Hello there,

I tried the StationaryProblem which defined below with NeuralNetworkReductor and the loss was 1.332e+28

```
problem = StationaryProblem(
    domain=RectDomain(domain=([0,0], [5, 5])),

    rhs=LincombFunction(
        [ExpressionFunction('10', 2), ConstantFunction(1., 2)],
        [ProjectionParameterFunctional('mu'), 0.1]),

    diffusion=LincombFunction(
        [ExpressionFunction('1 - x[0]', 2), ExpressionFunction('x[0]', 2)],
        [ProjectionParameterFunctional('mu'), 1]),

    dirichlet_data=LincombFunction(
        [ExpressionFunction('2 * x[0]', 2), ConstantFunction(1., 2)],
        [ProjectionParameterFunctional('mu'), 0.5]),

    name='2DProblem'
)
```

So I thought the input should be normalized, and  added the config

```
reductor = NeuralNetworkReductor(training_set=ts, validation_set=vs,
                                 l2_err=1e-5, ann_mse=1e-5)
reductor.scaling_parameters = {'min_inputs': 0, 'max_inputs': 10,
                                'min_targets': 0, 'max_targets': 10
                               }
rom = reductor.reduce(restarts=100)
```

And it failed at `Computing training samples`

```
13:27 NeuralNetworkReductor: Computing training samples ...
Output exceeds the [size limit](command:workbench.action.openSettings?%5B%22notebook.output.textLineLimit%22%5D). Open the full output data [in a text editor](command:workbench.action.openLargeOutput?dc1fd6d4-dbc1-4ece-bbeb-0ad5788e9ed7)
---------------------------------------------------------------------------
...

File [~/miniconda3/envs/dev/lib/python3.10/site-packages/pymor/reductors/neural_network.py:176](https://....vscode-cdn.net/home/dev/Codes/~/miniconda3/envs/dev/lib/python3.10/site-packages/pymor/reductors/neural_network.py:176), in NeuralNetworkReductor.reduce(self, hidden_layers, activation_function, optimizer, epochs, batch_size, learning_rate, loss_function, restarts, lr_scheduler, lr_scheduler_params, es_scheduler_params, weight_decay, log_loss_frequency)
    174 # build a reduced basis using POD and compute training data
    175 if not hasattr(self, 'training_data'):
--> 176     self.compute_training_data()
    178 layer_sizes = self._compute_layer_sizes(hidden_layers)
    180 # compute validation data
```

Did I misunderstand the normalization in NeuralNetworkReductor?

**Comment by HenKlei** [comment]
Hi @jacktang, great to hear that you are trying to use the neural network based reductors in pyMOR! The simplest way to enable scaling in the reductors (which is also the intended way to do it) is to set `scale_inputs=True` respectively `scale_outputs=True` when creating the reductor, see also [here](https://github.com/pymor/pymor/blob/2ebd506650690cfc9cc7573921ba10ffe4a9358f/src/pymor/reductors/neural_network.py#L92) and the lines below. In your case, something like
```
reductor = NeuralNetworkReductor(training_set=ts, validation_set=vs,
                                 l2_err=1e-5, ann_mse=1e-5,
                                 scale_inputs=True, scale_outputs=True)
```
should work. Let me know if you encounter further issues!

**Comment by jacktang** [comment]
Hello @HenKlei, thanks for your tip!

`scale_inputs=True` and `scale_outputs=True` solved the problem of "Computing training samples". But the mse loss of ANN is 2.530e+03. The error message:
```
NeuralNetworkTrainingFailed: Could not find neural network with prescribed loss of 1.000e-04 (best one found was 2.530e+03)!
```

And I paste whole code below:

```
problem = StationaryProblem(
    domain=RectDomain(domain=([0,0], [3, 3])),

    rhs=LincombFunction(
        [ExpressionFunction('10', 2), ConstantFunction(1., 2)],
        [ProjectionParameterFunctional('mu'), 0.1]),

    diffusion=LincombFunction(
        [ExpressionFunction('1 - x[0]', 2), ExpressionFunction('x[0]', 2)],
        [ProjectionParameterFunctional('mu'), 1]),

    dirichlet_data=LincombFunction(
        [ExpressionFunction('2 * x[0]', 2), ConstantFunction(1., 2)],
        [ProjectionParameterFunctional('mu'), 0.5]),

    name='2DProblem'
)

fom, _ = discretize_stationary_cg(problem, diameter=1/100)

parameter_space = fom.parameters.space((0.1, 1)) 

training_set = parameter_space.sample_uniformly(300)
validation_set = parameter_space.sample_randomly(100)

ts = []
vs = []

for mu in training_set:
    ts.append((mu, fom.solve(mu)))

for mu in validation_set:
    vs.append((mu, fom.solve(mu)))   

reductor = NeuralNetworkReductor(training_set=ts, validation_set=vs,
                                 l2_err=1e-5, ann_mse=1e-4, 
                                 #scale_inputs=True, scale_outputs=True
                                 )
rom = reductor.reduce(restarts=100)
```

 I think the "MOR with ANN" tutorial contains some "tricks":
1. Domain is in [0, 1] range
2. parameter space is in [0, 1] range

I've tested domain / parameter space bigger than 1, and the relative errors between built ROM and FOM is very large. What I missed?

**Reply by HenKlei** [reply]
Dear @jacktang, sorry for the late reply.

I just copied your code into a new file and ran it with some small changes (I reduced the number of training and validation parameters quite a bit...). With the code below, I get an average relative error on the test set of around 0.003 with a median speedup of 305 on my machine (the exact numbers will probably differ on your system). That sounds good to me. What do you think?

```
import time
import numpy as np

from pymor.basic import *
from pymor.reductors.neural_network import NeuralNetworkReductor

problem = StationaryProblem(
    domain=RectDomain(domain=([0, 0], [3, 3])),

    rhs=LincombFunction(
        [ExpressionFunction('10', 2), ConstantFunction(1., 2)],
        [ProjectionParameterFunctional('mu'), 0.1]),

    diffusion=LincombFunction(
        [ExpressionFunction('1 - x[0]', 2), ExpressionFunction('x[0]', 2)],
        [ProjectionParameterFunctional('mu'), 1]),

    dirichlet_data=LincombFunction(
        [ExpressionFunction('2 * x[0]', 2), ConstantFunction(1., 2)],
        [ProjectionParameterFunctional('mu'), 0.5]),

    name='2DProblem'
)

fom, _ = discretize_stationary_cg(problem, diameter=1/100)

parameter_space = fom.parameters.space((0.1, 1))

training_set = parameter_space.sample_uniformly(50)
validation_set = parameter_space.sample_randomly(10)
test_set = parameter_space.sample_randomly(10)

ts = []
vs = []

for mu in training_set:
    ts.append((mu, fom.solve(mu)))

for mu in validation_set:
    vs.append((mu, fom.solve(mu)))

reductor = NeuralNetworkReductor(training_set=ts, validation_set=vs,
                                 l2_err=1e-5, ann_mse=1e-4,
                                 scale_inputs=True, scale_outputs=True)
rom = reductor.reduce(restarts=100)

U = fom.solution_space.empty(reserve=len(test_set))
U_red = fom.solution_space.empty(reserve=len(test_set))
speedups = []

for mu in test_set:
    tic = time.perf_counter()
    U.append(fom.solve(mu))
    time_fom = time.perf_counter() - tic

    tic = time.perf_counter()
    U_red.append(reductor.reconstruct(rom.solve(mu)))
    time_rom = time.perf_counter() - tic

    speedups.append(time_fom / time_rom)

absolute_errors = (U - U_red).norm()
relative_errors = (U - U_red).norm() / U.norm()

print(f'Average absolute error: {np.average(absolute_errors)}')
print(f'Average relative error: {np.average(relative_errors)}')
print(f'Median of speedup: {np.median(speedups)}')
```

**Reply by jacktang** [reply]
It's wired. I can't reach `ann_mse=1e-4`, the minimum  of mse is `2.463e-03`.

the version of pymor  is 2022.2.0 and torch is 1.12.1. Should I update pymor to main branch?

**Reply by HenKlei** [reply]
That should not make too much of a difference, I think. There are of course many things influencing the outcome of the training, so adjusting the tolerances might be necessary. Maybe just setting `ann_mse=None` is also fine. This results in pyMOR selecting the neural network with the smallest loss after performing the prescribed number of restarts. Then you definitely get a reduced model.

**Reply by jacktang** [reply]
I think it matters :), However `ann_mse=None` finally let me get the ROM and relative errors is 0.02280900267497216. Can we make a conclusion that the larger domain size, the less accurate the ROM is?

**Reply by HenKlei** [reply]
Did you try with the pyMOR main branch version? You might be right that although nothing directly in the neural network-based reductors changed, the results could be different.

I think one cannot directly say that the accuracy correlates with the domain size or the discretization size. The ROM accuracy is bounded by the sum of two error terms: The error of the reduced basis (which is controlled (at least on the training set) by `l2_err`) and the loss of the neural network (which is also controlled on the training set by the MSE of the neural network). From what you reported above, the neural network training seems to be the bottleneck in your code so far. Maybe you can try to play around a bit with the size of the neural network (number of layers, neurons per layer, etc.) and try to improve the training results a bit. This should also lead to an improvement in the ROM accuracy.

By the way, what does the speedup say on your machine?

**Reply by jacktang** [reply]
I don't save the previous result, I will update it later.
Now I enlarge `parameter_space` from [0.1, 1] to [0.1, 3], mse loss: 0.0005279900431673213, and relative err: 0.01817494460448577, median speedup: 559.6899929454491.
I'd like to compare it with DMD in next days

**Reply by jacktang** [reply]
> By the way, what does the speedup say on your machine?

I re-run the case, and the result are list below:

relative errors: 0.02280900267497216
median speedup: 457.65886309864993
